server/command.go

```diff
@@ -79,40 +122,160 @@ func (p *Plugin) validateCommand(action string, parameters []string) string {
-				p.postCommandResponse(args, "error occurred while processing greeting for team `%s`: `%s`", teamName, err)
-				return
-			}
+func (p *Plugin) validatePreviewPrivileges(teamID string, args *model.CommandArgs) bool {
+	_, teamMemberErr := p.API.GetTeamMember(teamID, args.UserId)
+	if teamMemberErr != nil {
+		if teamMemberErr.StatusCode == 404 {
+			p.postCommandResponse(args, "You are not a member of that team.")
+			return false
+		}
+		p.postCommandResponse(args, "error occurred while getting the Team Admin Role `%s`: `%s`", teamID, teamMemberErr)
+		return false
+	}
+	doesUserHavePrivileges := p.checkForProperPrivileges(args, args.UserId, teamID)
+
+	return doesUserHavePrivileges
+}
+
+func (p *Plugin) checkForProperPrivileges(args *model.CommandArgs, userID string, teamID string) bool {
+	isSysadmin, sysAdminError := p.hasSysadminRole(userID)
+	isTeamAdmin, teamAdminError := p.hasTeamAdminRole(userID, teamID)
+
+	if sysAdminError != nil {
+		p.postCommandResponse(args, "error occurred while getting the System Admin Role `%s`: `%s`", teamID, sysAdminError)
+		return false
+	}
+	if teamAdminError != nil {
+		p.postCommandResponse(args, "error occurred while getting the Team Admin Role `%s`: `%s`", teamID, teamAdminError)
+		return false
+	}
+	if !isSysadmin && !isTeamAdmin {
+		p.postCommandResponse(args, "You do not have the proper privileges to control this Team's welcome messages.")
+		return false
+	}
+	return true
+}
+
+func (p *Plugin) showDynamicMessages(args *model.CommandArgs) bool {
+	// Checking dynamic welcome messages
+	teamsList, teamErr := p.API.GetTeams()
+	if teamErr != nil {
+		p.postCommandResponse(args, "Error occurred while getting list of teams: %s", teamErr)
+		return false
+	}
 
-			found = true
+	var dynamicTeamWelcome []string
+
+	for _, team := range teamsList {
+		key := fmt.Sprintf("%s%s", welcomebotTeamWelcomeKey, team.Id)
+		teamMessage, appErr := p.API.KVGet(key)
+		if appErr != nil {
+			p.postCommandResponse(args, "Error occurred while retrieving the welcome messages: %s", appErr)
+			return false
+		}
+		if teamMessage != nil {
+			dynamicTeamWelcome = append(dynamicTeamWelcome, team.DisplayName)
 		}
 	}
 
-	if !found {
-		p.postCommandResponse(args, "team `%s` has not been found", teamName)
+	if len(dynamicTeamWelcome) == 0 {
+		return false
 	}
+
+	var str strings.Builder
+	str.WriteString("Teams for which welcome messages are defined:")
+	for _, team := range dynamicTeamWelcome {
+		str.WriteString(fmt.Sprintf("\n * %s", team))
+	}
+	p.postCommandResponse(args, str.String())
+
+	return true
+}
+
+func (p *Plugin) executeCommandPreview(teamName string, args *model.CommandArgs) {
+	// Retrieve Team to check if a message already exists within the KV pair set
+	team, err := p.API.GetTeamByName(teamName)
+	if err != nil {
+		p.postCommandResponse(args, teamRetrievalErr, err)
+		return
+	}
+	teamID := team.Id
+
+	validPrivileges := p.validatePreviewPrivileges(teamID, args)
+	if !validPrivileges {
+		return
```

Comment:
No error for the user?

Reply:
The response to the user if they do not have preview privileges is found inside the function `validatePreviewPrivileges`. That is where we check what kind of permissions error it is before returning it back to the user.

Reply:
Similar as in a comment above I think the signature